core/src/main/java/com/graphhopper/util/ShallowImmutablePointList.java

```diff
@@ -0,0 +1,205 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.util;
+
+import java.util.List;
+
+/**
+ * This is a shallow copy of a PointList. This class can be used to create a memory efficient copy of a PointList.
+ * You have to be aware, that if the wrapped PointList changes, the shallow copy will change as well. This can lead to
+ * unexpected results. We recommend making the wrapped PointList immutable {@link PointList#makeImmutable()}.
+ *
+ * @author Robin Boldt
+ */
+public class ShallowImmutablePointList extends PointList {
+
+    private final static String IMMUTABLE_ERR = "This class is immutable, you are not allowed to change it";
+
+    private final int fromOffset;
+    private final int toOffset;
+    private final PointList wrappedPointList;
+
+    public ShallowImmutablePointList(int fromOffset, int toOffset, PointList wrappedPointList) {
+        if (fromOffset > toOffset)
+            throw new IllegalArgumentException("from must be smaller or equal to end");
+        if (fromOffset < 0 || toOffset > wrappedPointList.getSize())
+            throw new IllegalArgumentException("Illegal interval: " + fromOffset + ", " + toOffset);
+        this.fromOffset = fromOffset;
+        this.toOffset = toOffset;
+        this.wrappedPointList = wrappedPointList;
+    }
+
+    @Override
+    public int size() {
+        return toOffset - fromOffset;
+    }
+
+    @Override
+    public int getSize() {
+        return size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    @Override
+    public double getLatitude(int index) {
+        if (index > getSize())
+            throw new ArrayIndexOutOfBoundsException(ERR_MSG + " index:" + index + ", size:" + getSize());
+        return wrappedPointList.getLatitude(fromOffset + index);
+    }
+
+    @Override
+    public double getLongitude(int index) {
+        if (index > getSize())
+            throw new ArrayIndexOutOfBoundsException(ERR_MSG + " index:" + index + ", size:" + getSize());
+        return wrappedPointList.getLongitude(fromOffset + index);
+    }
+
+    @Override
+    public double getElevation(int index) {
+        if (index > getSize())
+            throw new ArrayIndexOutOfBoundsException(ERR_MSG + " index:" + index + ", size:" + getSize());
+        return wrappedPointList.getElevation(fromOffset + index);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < getSize(); i++) {
+            if (i > 0)
+                sb.append(", ");
+
+            sb.append('(');
+            sb.append(getLatitude(i));
+            sb.append(',');
+            sb.append(getLongitude(i));
+            if (is3D) {
+                sb.append(',');
+                sb.append(getElevation(i));
+            }
+            sb.append(')');
+        }
+        return sb.toString();
```

Comment:
Is this duplication necessary?

Reply:
I just spotted an issue here, thanks. The method should run from `from` to `end` and not from 0 to size. I can try to reduce the code duplicates a bit :).